sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/RampupThrottlingFn.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.datastore;
+
+import java.io.IOException;
+import java.io.Serializable;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Sum;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.util.BackOff;
+import org.apache.beam.sdk.util.FluentBackoff;
+import org.apache.beam.sdk.util.MovingFunction;
+import org.apache.beam.sdk.util.Sleeper;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An implementation of a client-side throttler that enforces a gradual ramp-up, broadly in line
+ * with Datastore best practices. See also
+ * https://cloud.google.com/datastore/docs/best-practices#ramping_up_traffic.
+ */
+public class RampupThrottlingFn<T> extends DoFn<T, T> implements Serializable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RampupThrottlingFn.class);
+  private static final double BASE_BUDGET = 500.0;
+  private static final Duration RAMP_UP_INTERVAL = Duration.standardMinutes(5);
+  private static final FluentBackoff fluentBackoff = FluentBackoff.DEFAULT;
+
+  private final int numWorkers;
+  private final Counter throttlingMsecs =
+      Metrics.counter(RampupThrottlingFn.class, "throttling-msecs");
+
+  // Initialized on Beam setup.
```

Comment:
This is something I actually implemented differently in Java accidentally, I've changed it back now.  
So I was going by [this diagram](https://beam.apache.org/images/dofn-sequence-diagram.svg) from the Beam docs. As far as I know, there is an initial DoFn instance with the `firstInstant` set, and that will get cloned to each worker. Since `firstInstant` is serializable, it stays the same, while the moving function with a per-second budget has to be reinitialized, so the budget may be out-of-date for up to a second when a worker gets created or restarted.  
However, I was reinitializing `firstInstant` when Beam calls `setup()`, so workers would actually start over completely. I've removed that, so now it should match the Python version and get the right throughput.

Reply:
Thanks for fixing the bug. Yeah `SetUp` is called per DoFn instance. Now that `firstInstant` is initialized only once in ctor, could you also update the comment accordingly?

Reply:
Done.